File: apps/ios/Sources/IdentityStore.swift
import Foundation
import Security

/// Persists the Rust FFI identity **seed** (32 bytes) in the iOS Keychain so this device
/// keeps a **stable `device_id`** across launches. Without this the `MobileClient`
/// generates a fresh identity each launch, so the desktop's trust of the phone (and thus
/// pairing) would not survive an app restart. The seed is private key material, hence the
/// Keychain rather than `UserDefaults`.
enum IdentityStore {
    private static let service = "ai.unlikeother.mouser"
    private static let account = "identity-seed"

    /// The persisted seed, or `nil` on first run / if it was never saved.
    static func load() -> Data? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        guard status == errSecSuccess, let data = item as? Data, !data.isEmpty else {
            return nil
        }
        return data
    }

    /// Store (upsert) the seed. Delete-then-add keeps it a simple idempotent write.
    static func save(_ seed: Data) {
        let base: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(base as CFDictionary)
        var add = base
        add[kSecValueData as String] = seed
        add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlock
        SecItemAdd(add as CFDictionary, nil)
    }
}

File: apps/android/app/src/main/kotlin/ai/unlikeother/mouser/companion/IdentityStore.kt
package ai.unlikeother.mouser.companion

import android.content.Context
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey

/**
 * Persists the Rust FFI identity **seed** (32 bytes) in Android secure storage so this
 * device keeps a **stable `device_id`** across launches. Android parity of the iOS
 * `IdentityStore` (which uses the Keychain).
 *
 * Without this the [uniffi.mouser_ffi.MobileClient] generates a fresh identity each
 * launch, so the desktop's trust of the phone (and thus pairing) would not survive an
 * app restart. The seed is private key material, hence [EncryptedSharedPreferences]
 * (AES-256-GCM, wrapped by a Keystore-backed [MasterKey]) rather than a plain
 * `SharedPreferences`.
 *
 * The seed is stored Base64-encoded because [EncryptedSharedPreferences] only persists
 * strings. All reads/writes are best-effort: if the Keystore is unavailable (e.g. a
 * device with a broken keystore provider) we log and fall back to a fresh identity
 * rather than crash — the only cost is re-pairing.
 */
class IdentityStore(context: Context) {

    private val appContext = context.applicationContext

    /** Lazily-built encrypted prefs; null if the Keystore-backed store can't be opened. */
    private val prefs by lazy { openPrefs() }

    /** The persisted seed, or `null` on first run / if it was never saved / unreadable. */
    fun load(): ByteArray? {
        val encoded = runCatching { prefs?.getString(KEY_SEED, null) }
            .onFailure { Log.w(TAG, "identity load failed", it) }
            .getOrNull() ?: return null
        return runCatching { android.util.Base64.decode(encoded, android.util.Base64.NO_WRAP) }
            .getOrNull()
            ?.takeIf { it.isNotEmpty() }
    }

    /** Store (upsert) the seed. Best-effort; a failure leaves the next launch to re-pair. */
    fun save(seed: ByteArray) {
        val encoded = android.util.Base64.encodeToString(seed, android.util.Base64.NO_WRAP)
        runCatching { prefs?.edit()?.putString(KEY_SEED, encoded)?.apply() }
            .onFailure { Log.w(TAG, "identity save failed", it) }
    }

    private fun openPrefs() = runCatching {
        val masterKey = MasterKey.Builder(appContext)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        EncryptedSharedPreferences.create(
            appContext,
            PREFS_NAME,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }.onFailure { Log.w(TAG, "EncryptedSharedPreferences unavailable", it) }.getOrNull()

    private companion object {
        const val TAG = "IdentityStore"
        const val PREFS_NAME = "mouser.identity"
        const val KEY_SEED = "identity-seed"
    }
}
